pipeline/pipeline.go

```diff
@@ -0,0 +1,111 @@
+package pipeline
+
+import "github.com/51Degrees/device-detection-go/v4/dd"
+
+type Pipeline struct {
+	logger                        logWrapper
+	dataFile                      string
+	licenceKey                    string
+	createTempDataCopy            bool
+	dataUpdateOnStartEnabled      bool
+	dataFileUrl                   string
+	dataFilePullEveryMs           int
+	isScheduledFilePullingEnabled bool
+	loggerEnabled                 bool
+	manager                       *dd.ResourceManager
+	config                        *dd.ConfigHash
+}
+
+// NewResultsHash creates a new ResultsHash object
+// uaCapacity is the initial capacity of the User-Agent cache
+// overridesCapacity is the initial capacity of the overrides cache
+func (p *Pipeline) NewResultsHash(uaCapacity uint32, overridesCapacity uint32) *dd.ResultsHash {
+	return dd.NewResultsHash(p.manager, uaCapacity, overridesCapacity)
+}
+
+// Run starts the pipeline
+func (p *Pipeline) Run() error {
+	dataFile, err := dd.GetFilePath(".", []string{p.dataFile})
+	if err != nil {
+		return err
+	}
+	p.dataFile = dataFile
+
+	// TODO: what are properties of dd.ResourceManager?
+	err = dd.InitManagerFromFile(p.manager, *p.config, "", p.dataFile)
+	if err != nil {
+		return err
+	}
+
+	if p.isScheduledFilePullingEnabled {
+		go p.scheduleFilePulling()
+	}
+
+	return nil
+}
+
+type PipelineOptions func(cfg *Pipeline) error
+
+// what should this do? c# has it and it's used in the example.
+func UseOnPremise(dataFile, licenceKey string, createTempDataCopy bool) PipelineOptions {
+
+	return func(cfg *Pipeline) error {
+		cfg.dataFile = dataFile
+		cfg.licenceKey = licenceKey
+		cfg.createTempDataCopy = createTempDataCopy
+
+		return nil
+	}
+}
+
+// WithDataUpdateUrl sets the URL to pull data from and the interval in milliseconds
+func WithDataUpdateUrl(url string, everyMs int) PipelineOptions {
```

Comment:
So various `With...` methods is a pattern of .NET Builder-objects that are building an instance of other object - that actually does the job..  In go - I am not sure there is Builder-pattern.  If it is ok with go then maybe let's use it otherwise - maybe it's easier just to create a DeviceDetectionOnPremiseEngineConfiguration structure right away without various With method calls?

Reply:
https://dev.to/kittipat1413/understanding-the-options-pattern-in-go-390c